### LSE Data Analytics Online Career Accelerator

# Course 2: Data Analytics using Python

## Assignment: Diagnostic Analysis using Python

You’ll be working with real-world data to address a problem faced by the National Health Service (NHS). The analysis will require you to utilise Python to explore the available data, create visualisations to identify trends, and extract meaningful insights to inform decision-making. 

### A note for students using this template
This Jupyter Notebook is a template you can use to complete the Course 2 assignment: Diagnostic Analysis using Python. 

Keep in mind: 
- You are **not required** to use this template to complete the assignment. 
- If you decide to use this template for your assignment, make a copy of the notebook and save it using the assignment naming convention: **LastName_FirstName_DA201_Assignment_Notebook.ipynb**.
- The workflow suggested in this template follows the Assignment Activities throughout the course.
- Refer to the guidance on the Assignment Activity pages for specific details. 
- The markup and comments in this template identify the key elements you need to complete before submitting the assignment.
- Make this notebook your own by adding your process notes and rationale using markdown, add links, screenshots, or images to support your analysis, refine or clarify the comments, and change the workflow to suit your process.
- All elements should be functional and visible in your Notebook. 
- Be sure to push your notebook to GitHub after completing each Assignment Activity.

 > ***Markdown*** Remember to change cell types to `Markdown`. You can review [Markdown basics](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) to find out how to add formatted text, links, and images to your notebook.

# 

# Assignment activity 1

### Insert proof of your GitHub repository. This can be a link or screenshot showing your repo.

In [1]:
# My GitHub repository.
# https://github.com/angjiashin/LSE_DA_NHS_analysis

![Repository](https://raw.githubusercontent.com/angjiashin/LSE_DA_NHS_analysis/main/images/github_repo.png)

# 

# Assignment activity 2

How many locations are there in the data set?

What are the five locations with the highest number of records?

How many service settings, context types, national categories, and appointment statuses are there?

### Prepare your workstation

In [1]:
# Import the necessary libraries.
import pandas as pd
import numpy as np

# Optional - Ignore warnings.
import warnings
warnings.filterwarnings('ignore')

# Run all nodes interactively.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
# Import and sense-check the actual_duration.csv data set as ad.
ad = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/actual_duration.csv')

# View the DataFrame.
ad.shape
ad.head(5)
ad.tail(5)

(137793, 8)

sub_icb_location_code sub_icb_location_ons_code  \
0                   00L                 E38000130   
1                   00L                 E38000130   
2                   00L                 E38000130   
3                   00L                 E38000130   
4                   00L                 E38000130   

                        sub_icb_location_name icb_ons_code region_ons_code  \
0  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
1  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
2  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
3  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   
4  NHS North East and North Cumbria ICB - 00L    E54000050       E40000012   

  appointment_date         actual_duration  count_of_appointments  
0        01-Dec-21           31-60 Minutes                    364  
1        01-Dec-21           21-30 Minutes                    619  
2        01-Dec-21            6-10 Minutes                   1698  
3        01-Dec-21  Unknown / Data Quality                   1277  
4        01-Dec-21           16-20 Minutes                    730

sub_icb_location_code sub_icb_location_ons_code  \
137788                 X2C4Y                 E38000254   
137789                 X2C4Y                 E38000254   
137790                 X2C4Y                 E38000254   
137791                 X2C4Y                 E38000254   
137792                 X2C4Y                 E38000254   

                 sub_icb_location_name icb_ons_code region_ons_code  \
137788  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137789  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137790  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137791  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   
137792  NHS West Yorkshire ICB - X2C4Y    E54000054       E40000012   

       appointment_date actual_duration  count_of_appointments  
137788        30-Jun-22   31-60 Minutes                    430  
137789        30-Jun-22   21-30 Minutes                    751  
137790        30-Jun-22   16-20 Minutes                    921  
137791        30-Jun-22   11-15 Minutes                   1439  
137792        30-Jun-22     1-5 Minutes                   1739

The actual_duration data set has 137,783 rows and 8 columns.

In [3]:
# Determine whether there are missing values.
ad.isnull().sum().sort_values(ascending = False)

sub_icb_location_code        0
sub_icb_location_ons_code    0
sub_icb_location_name        0
icb_ons_code                 0
region_ons_code              0
appointment_date             0
actual_duration              0
count_of_appointments        0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [4]:
# Determine the metadata of the data set.
ad.info()
print(ad.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137793 entries, 0 to 137792
Data columns (total 8 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   sub_icb_location_code      137793 non-null  object
 1   sub_icb_location_ons_code  137793 non-null  object
 2   sub_icb_location_name      137793 non-null  object
 3   icb_ons_code               137793 non-null  object
 4   region_ons_code            137793 non-null  object
 5   appointment_date           137793 non-null  object
 6   actual_duration            137793 non-null  object
 7   count_of_appointments      137793 non-null  int64 
dtypes: int64(1), object(7)
memory usage: 8.4+ MB
['sub_icb_location_code', 'sub_icb_location_ons_code', 'sub_icb_location_name', 'icb_ons_code', 'region_ons_code', 'appointment_date', 'actual_duration', 'count_of_appointments']


The actual_duration data set consist of 2 data types: object 
('sub_icb_location_code', 'sub_icb_location_ons_code', 
'sub_icb_location_name', 'icb_ons_code', 'region_ons_code', 'appointment_date', 'actual_duration') 
and integer.('count_of_appointments')

In [5]:
# Determine the descriptive statistics of the data set.
ad.describe().T

count         mean          std  min    25%    50%  \
count_of_appointments  137793.0  1219.080011  1546.902956  1.0  194.0  696.0   

                          75%      max  
count_of_appointments  1621.0  15400.0

In [6]:
# Import and sense-check the appointments_regional.csv data set as ar.
ar = pd.read_csv('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/appointments_regional.csv') 

# View the DataFrame.
ar.shape
ar.head(5)
ar.tail(5)

(596821, 7)

icb_ons_code appointment_month appointment_status hcp_type appointment_mode  \
0    E54000034           2020-01           Attended       GP     Face-to-Face   
1    E54000034           2020-01           Attended       GP     Face-to-Face   
2    E54000034           2020-01           Attended       GP     Face-to-Face   
3    E54000034           2020-01           Attended       GP     Face-to-Face   
4    E54000034           2020-01           Attended       GP     Face-to-Face   

  time_between_book_and_appointment  count_of_appointments  
0                             1 Day                   8107  
1                    15  to 21 Days                   6791  
2                       2 to 7 Days                  20686  
3                    22  to 28 Days                   4268  
4                     8  to 14 Days                  11971

icb_ons_code appointment_month appointment_status hcp_type  \
596816    E54000050           2022-06            Unknown  Unknown   
596817    E54000050           2022-06            Unknown  Unknown   
596818    E54000050           2022-06            Unknown  Unknown   
596819    E54000050           2022-06            Unknown  Unknown   
596820    E54000050           2022-06            Unknown  Unknown   

       appointment_mode time_between_book_and_appointment  \
596816          Unknown                       2 to 7 Days   
596817          Unknown                    22  to 28 Days   
596818          Unknown                     8  to 14 Days   
596819          Unknown                 More than 28 Days   
596820          Unknown                          Same Day   

        count_of_appointments  
596816                     21  
596817                      8  
596818                     28  
596819                     17  
596820                     10

The appointments_regional data set has 596,821 rows and 7 columns.

In [7]:
# Determine whether there are missing values.
ar.isnull().sum().sort_values(ascending = False)

icb_ons_code                         0
appointment_month                    0
appointment_status                   0
hcp_type                             0
appointment_mode                     0
time_between_book_and_appointment    0
count_of_appointments                0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [8]:
# Determine the metadata of the data set.
ar.info()
print(ar.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 596821 entries, 0 to 596820
Data columns (total 7 columns):
 #   Column                             Non-Null Count   Dtype 
---  ------                             --------------   ----- 
 0   icb_ons_code                       596821 non-null  object
 1   appointment_month                  596821 non-null  object
 2   appointment_status                 596821 non-null  object
 3   hcp_type                           596821 non-null  object
 4   appointment_mode                   596821 non-null  object
 5   time_between_book_and_appointment  596821 non-null  object
 6   count_of_appointments              596821 non-null  int64 
dtypes: int64(1), object(6)
memory usage: 31.9+ MB
['icb_ons_code', 'appointment_month', 'appointment_status', 'hcp_type', 'appointment_mode', 'time_between_book_and_appointment', 'count_of_appointments']


The appointments_regional data set consist of 2 data types: object 
('icb_ons_code', 'appointment_month', 
'appointment_status', 'hcp_type', 'appointment_mode', 'time_between_book_and_appointment') 
and integer.('count_of_appointments')

In [9]:
# Determine the descriptive statistics of the data set.
ar.describe().T

count         mean          std  min  25%   50%  \
count_of_appointments  596821.0  1244.601857  5856.887042  1.0  7.0  47.0   

                         75%       max  
count_of_appointments  308.0  211265.0

In [10]:
# Import and sense-check the national_categories.xlsx data set as nc.
nc = pd.read_excel('/Users/shin/Desktop/lse_daca/course_2/LSE_DA201_Assignment_files/national_categories.xlsx') 

# View the DataFrame.
nc.shape
nc.head(5)
nc.tail(5)

(817394, 8)

appointment_date icb_ons_code                       sub_icb_location_name  \
0       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
1       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
2       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
3       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
4       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   

        service_setting            context_type  \
0  Primary Care Network  Care Related Encounter   
1                 Other  Care Related Encounter   
2      General Practice  Care Related Encounter   
3      General Practice  Care Related Encounter   
4      General Practice  Care Related Encounter   

                        national_category  count_of_appointments  \
0  Patient contact during Care Home Round                      3   
1                         Planned Clinics                      7   
2                              Home Visit                     79   
3              General Consultation Acute                    725   
4            Structured Medication Review                      2   

  appointment_month  
0           2021-08  
1           2021-08  
2           2021-08  
3           2021-08  
4           2021-08

appointment_date icb_ons_code           sub_icb_location_name  \
817389       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817390       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817391       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817392       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   
817393       2022-06-30    E54000054  NHS West Yorkshire ICB - X2C4Y   

                  service_setting            context_type  \
817389  Extended Access Provision  Care Related Encounter   
817390  Extended Access Provision  Care Related Encounter   
817391  Extended Access Provision  Care Related Encounter   
817392  Extended Access Provision  Care Related Encounter   
817393  Extended Access Provision  Care Related Encounter   

                   national_category  count_of_appointments appointment_month  
817389   Unplanned Clinical Activity                     12           2022-06  
817390               Planned Clinics                      4           2022-06  
817391    Planned Clinical Procedure                     92           2022-06  
817392  General Consultation Routine                      4           2022-06  
817393    General Consultation Acute                     19           2022-06

The national_categories data set has 817,394 rows and 8 columns.

In [11]:
# Determine whether there are missing values.
nc.isnull().sum().sort_values(ascending = False)

appointment_date         0
icb_ons_code             0
sub_icb_location_name    0
service_setting          0
context_type             0
national_category        0
count_of_appointments    0
appointment_month        0
dtype: int64

Based on the shape and null value counts, it appears that there are no missing or null values.

In [12]:
# Determine the metadata of the data set.
nc.info()
print(nc.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB
['appointment_date', 'icb_ons_code', 'sub_icb_location_name', 'service_setting', 'context_type', 'national_category', 'count_of_appointments', 'appointment_month']


The appointments_regional data set consist of 3 data types:- 

1) object: 'appointment_date', 'icb_ons_code', 'sub_icb_location_name', 'service_setting', 'context_type', 'national_category', 'appointment_month'

2) integer: 'count_of_appointments'

3) datetime64[ns]: 'appointment_date'

In [13]:
# Determine the descriptive statistics of the data set.
nc.describe().T

count        mean        std  min  25%   50%    75%  \
count_of_appointments  817394.0  362.183684  1084.5766  1.0  7.0  25.0  128.0   

                           max  
count_of_appointments  16590.0

### Explore the data set

**Question 1:** How many locations are there in the data set?

In [14]:
# Determine the number of locations.
print(nc["sub_icb_location_name"].value_counts())
location_count = nc['sub_icb_location_name'].nunique()
print(f'The number of locations in the data set is: {location_count}.')

NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
                                               ...  
NHS North East and North Cumbria ICB - 00N      4210
NHS Lancashire and South Cumbria ICB - 02G      4169
NHS Cheshire and Merseyside ICB - 01V           3496
NHS Cheshire and Merseyside ICB - 01T           3242
NHS Greater Manchester ICB - 00V                2170
Name: sub_icb_location_name, Length: 106, dtype: int64
The number of locations in the data set is: 106.


**Question 2:** What are the five locations with the highest number of records?



In [15]:
# Determine the top five locations based on record count.
print(f'The 5 locations with the highest number of records are: ')
print(nc['sub_icb_location_name'].value_counts().sort_values(ascending=False).head())

The 5 locations with the highest number of records are: 
NHS North West London ICB - W2U3Z              13007
NHS Kent and Medway ICB - 91Q                  12637
NHS Devon ICB - 15N                            12526
NHS Hampshire and Isle Of Wight ICB - D9Y0V    12171
NHS North East London ICB - A3A8R              11837
Name: sub_icb_location_name, dtype: int64


**Question 3:** How many service settings, context types, national categories, and appointment statuses are there?

In [16]:
# Determine the number of service settings.
print(nc["service_setting"].value_counts())
service_setting_count = nc["service_setting"].value_counts().count()
print(f' There are {service_setting_count} service settings.')

General Practice             359274
Primary Care Network         183790
Other                        138789
Extended Access Provision    108122
Unmapped                      27419
Name: service_setting, dtype: int64
 There are 5 service settings.


In [17]:
# Determine the number of context types.
print(nc["context_type"].value_counts())
context_type_count = nc["context_type"].value_counts().count()
print(f' There are {context_type_count} context types.')

Care Related Encounter    700481
Inconsistent Mapping       89494
Unmapped                   27419
Name: context_type, dtype: int64
 There are 3 context types.


In [18]:
# Determine the number of national categories.
print(nc["national_category"].value_counts())
national_category_count = nc["national_category"].value_counts().count()
print(f' There are {national_category_count} national categories.')

Inconsistent Mapping                                                   89494
General Consultation Routine                                           89329
General Consultation Acute                                             84874
Planned Clinics                                                        76429
Clinical Triage                                                        74539
Planned Clinical Procedure                                             59631
Structured Medication Review                                           44467
Service provided by organisation external to the practice              43095
Home Visit                                                             41850
Unplanned Clinical Activity                                            40415
Patient contact during Care Home Round                                 28795
Unmapped                                                               27419
Care Home Visit                                                        26644

In [19]:
# Determine the number of appointment status.
print(ar["appointment_status"].value_counts())
appointment_status_count = ar["appointment_status"].value_counts().count()
print(f' There are {appointment_status_count} appointment statuses.')

Attended    232137
Unknown     201324
DNA         163360
Name: appointment_status, dtype: int64
 There are 3 appointment statuses.


# 

# Assignment activity 3

### Continue to explore the data and search for answers to more specific questions posed by the NHS.

**Question 1:** Between what dates were appointments scheduled? 

In [20]:
# View the first five rows of appointment_date for the ad DataFrame to determine the date format.
ad['appointment_date'].head()

0    01-Dec-21
1    01-Dec-21
2    01-Dec-21
3    01-Dec-21
4    01-Dec-21
Name: appointment_date, dtype: object

In [21]:
# View the first five rows of appointment_date for the nc DataFrame to determine the date format.
nc['appointment_date'].head()

0   2021-08-02
1   2021-08-02
2   2021-08-02
3   2021-08-02
4   2021-08-02
Name: appointment_date, dtype: datetime64[ns]

In [22]:
# Change the date format of ad['appointment_date'].
ad['appointment_date'] = pd.to_datetime(ad.appointment_date)

# View the DateFrame.
ad['appointment_date'].head()

0   2021-12-01
1   2021-12-01
2   2021-12-01
3   2021-12-01
4   2021-12-01
Name: appointment_date, dtype: datetime64[ns]

In [23]:
# Change the date format of nc['appointment_date'].
nc['appointment_date'] = pd.to_datetime(nc.appointment_date)

# View the DateFrame.
nc['appointment_date'].head()

0   2021-08-02
1   2021-08-02
2   2021-08-02
3   2021-08-02
4   2021-08-02
Name: appointment_date, dtype: datetime64[ns]

In [24]:
# Determine the minimum and maximum dates in the ad DataFrame.
# Use appropriate docstrings.
min_date = ad['appointment_date'].min()
max_date = ad['appointment_date'].max()
print(f'First date of scheduled appointments: {min_date}')
print(f'Last date of scheduled appointments: {max_date}')      

First date of scheduled appointments: 2021-12-01 00:00:00
Last date of scheduled appointments: 2022-06-30 00:00:00


In [25]:
# Determine the minimum and maximum dates in the nc DataFrame.
# Use appropriate docstrings.
min_date = nc['appointment_date'].min()
max_date = nc['appointment_date'].max()
print(f'First date of scheduled appointments: {min_date}')
print(f'Last date of scheduled appointments: {max_date}')      

First date of scheduled appointments: 2021-08-01 00:00:00
Last date of scheduled appointments: 2022-06-30 00:00:00


**Question 2:** Which service setting was the most popular for NHS North West London from 1 January to 1 June 2022?

In [26]:
# For each of these service settings, determine the number of records available for the period and the location. 
nc_subset = nc.loc[(nc['sub_icb_location_name'] == 'NHS North West London ICB - W2U3Z') & 
                   (nc['appointment_date'] >= '2022-01-01') & 
                   (nc['appointment_date'] <= '2022-06-01')]

# View the output.
print(nc_subset["service_setting"].value_counts().sort_values(ascending = False).head(1))

General Practice    2104
Name: service_setting, dtype: int64


**Question 3:** Which month had the highest number of appointments?

In [27]:
nc.head()

appointment_date icb_ons_code                       sub_icb_location_name  \
0       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
1       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
2       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
3       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   
4       2021-08-02    E54000050  NHS North East and North Cumbria ICB - 00L   

        service_setting            context_type  \
0  Primary Care Network  Care Related Encounter   
1                 Other  Care Related Encounter   
2      General Practice  Care Related Encounter   
3      General Practice  Care Related Encounter   
4      General Practice  Care Related Encounter   

                        national_category  count_of_appointments  \
0  Patient contact during Care Home Round                      3   
1                         Planned Clinics                      7   
2                              Home Visit                     79   
3              General Consultation Acute                    725   
4            Structured Medication Review                      2   

  appointment_month  
0           2021-08  
1           2021-08  
2           2021-08  
3           2021-08  
4           2021-08

In [28]:
# Number of appointments per month == sum of count_of_appointments by month.
# Use the groupby() and sort_values() functions.
nc.groupby([
    nc['appointment_date'].dt.year.rename('year'), 
    nc['appointment_date'].dt.month.rename('month')])[['count_of_appointments']].sum().sort_values(by=
                                                                                                   'count_of_appointments',
                                                                                                    ascending = False).reset_index()

year  month  count_of_appointments
0   2021     11               30405070
1   2021     10               30303834
2   2022      3               29595038
3   2021      9               28522501
4   2022      5               27495508
5   2022      6               25828078
6   2022      1               25635474
7   2022      2               25355260
8   2021     12               25140776
9   2022      4               23913060
10  2021      8               23852171

From the dataframe above, November 2021 had the highest number of appointments.

**Question 4:** What was the total number of records per month?

In [29]:
# Total number of records per month.
nc.groupby('appointment_month')[['count_of_appointments']].sum().reset_index()

appointment_month  count_of_appointments
0            2021-08               23852171
1            2021-09               28522501
2            2021-10               30303834
3            2021-11               30405070
4            2021-12               25140776
5            2022-01               25635474
6            2022-02               25355260
7            2022-03               29595038
8            2022-04               23913060
9            2022-05               27495508
10           2022-06               25828078

# 

# Assignment activity 4

### Create visualisations and identify possible monthly and seasonal trends in the data.

In [30]:
# Import the necessary libraries.
# using plotly express as the main library for visualisation.
import plotly.express as px

### Objective 1
Create three visualisations indicating the number of appointments per month for service settings, context types, and national categories.

In [35]:
# Change the data type of the appointment month to string to allow for easier plotting.
nc['appointment_month'].astype(str)

0         2021-08
1         2021-08
2         2021-08
3         2021-08
4         2021-08
           ...   
817389    2022-06
817390    2022-06
817391    2022-06
817392    2022-06
817393    2022-06
Name: appointment_month, Length: 817394, dtype: object

In [58]:
# Aggregate on monthly level and determine the sum of records per month.
# create a seperate data set for each column.
# (service settings, context types, and national categories)
nc_ss = nc.groupby(['appointment_month', 'service_setting'])[['count_of_appointments']].sum().reset_index()
nc_ct = nc.groupby(['appointment_month', 'context_type'])[['count_of_appointments']].sum().reset_index()
nc_nc = nc.groupby(['appointment_month', 'national_category'])[['count_of_appointments']].sum().reset_index()

# View output.
nc_ss.head()
nc_ct.head()
nc_nc.head()

appointment_month            service_setting  count_of_appointments
0           2021-08  Extended Access Provision                 160927
1           2021-08           General Practice               21575852
2           2021-08                      Other                 449101
3           2021-08       Primary Care Network                 432448
4           2021-08                   Unmapped                1233843

appointment_month            context_type  count_of_appointments
0           2021-08  Care Related Encounter               20255235
1           2021-08    Inconsistent Mapping                2363093
2           2021-08                Unmapped                1233843
3           2021-09  Care Related Encounter               24404251
4           2021-09    Inconsistent Mapping                2782135

appointment_month                                  national_category  \
0           2021-08  Care Home Needs Assessment & Personalised Care...   
1           2021-08                                    Care Home Visit   
2           2021-08                                    Clinical Triage   
3           2021-08                         General Consultation Acute   
4           2021-08                       General Consultation Routine   

   count_of_appointments  
0                  29676  
1                  47583  
2                3704207  
3                4280920  
4                7756045

**Service settings:**

In [82]:
# Plot the appointments over the available date range, and review the service settings for months.
# Create a lineplot.
fig = px.line(nc_ss, x = 'appointment_month', y = 'count_of_appointments', color = 'service_setting',
                 title = 'Total appointment across months and service settings', 
                 labels = {'appointment_month': 'Month',
                           'count_of_appointments': 'Number of appointments',
                           'service_setting': 'Service Setting'})
fig.show()

**Context types:**

In [81]:
# Plot the appointments over the available date range, and review the context types for months.
# Create a lineplot.
fig = px.line(nc_ct, x = 'appointment_month', y = 'count_of_appointments', color = 'context_type',
                 title = 'Total appointment across months and context types', 
                 labels = {'appointment_month': 'Month',
                           'count_of_appointments': 'Number of appointments',
                           'context_type': 'Context Type'})
fig.show()

**National categories:**

In [80]:
# Plot the appointments over the available date range, and review the national categories for months.
# Create a lineplot.
fig = px.line(nc_nc, x = 'appointment_month', y = 'count_of_appointments', color = 'national_category',
                 title = 'Total appointment across months and national categories', 
                 labels = {'appointment_month': 'Month',
                           'count_of_appointments': 'Number of appointments',
                           'national_category': 'National Category'})
fig.show()

### Objective 2
Create four visualisations indicating the number of appointments for service setting per season. The seasons are summer (August 2021), autumn (October 2021), winter (January 2022), and spring (April 2022).

**Summer (August 2021):**

In [106]:
nc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 817394 entries, 0 to 817393
Data columns (total 8 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   appointment_date       817394 non-null  datetime64[ns]
 1   icb_ons_code           817394 non-null  object        
 2   sub_icb_location_name  817394 non-null  object        
 3   service_setting        817394 non-null  object        
 4   context_type           817394 non-null  object        
 5   national_category      817394 non-null  object        
 6   count_of_appointments  817394 non-null  int64         
 7   appointment_month      817394 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 49.9+ MB


In [109]:
# Create a separate data set that can be used in future weeks. 

nc_ss_day = nc.groupby(['appointment_date',
                        'appointment_month',
                        'service_setting'])[['count_of_appointments']].sum().reset_index()

nc_ss_summer = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2021-08']

nc_ss_autumn = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2021-10']

nc_ss_winter = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2022-01']

nc_ss_spring = nc_ss_day.loc[nc_ss_day['appointment_month'] == '2022-04']

# View output.
nc_ss_summer
nc_ss_autumn
nc_ss_winter
nc_ss_spring

appointment_date appointment_month            service_setting  \
0         2021-08-01           2021-08  Extended Access Provision   
1         2021-08-01           2021-08           General Practice   
2         2021-08-01           2021-08                      Other   
3         2021-08-01           2021-08       Primary Care Network   
4         2021-08-01           2021-08                   Unmapped   
..               ...               ...                        ...   
150       2021-08-31           2021-08  Extended Access Provision   
151       2021-08-31           2021-08           General Practice   
152       2021-08-31           2021-08                      Other   
153       2021-08-31           2021-08       Primary Care Network   
154       2021-08-31           2021-08                   Unmapped   

     count_of_appointments  
0                      438  
1                     3411  
2                      401  
3                      323  
4                     1054  
..                     ...  
150                   8281  
151                1125584  
152                  23181  
153                  21789  
154                  62380  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
305       2021-10-01           2021-10  Extended Access Provision   
306       2021-10-01           2021-10           General Practice   
307       2021-10-01           2021-10                      Other   
308       2021-10-01           2021-10       Primary Care Network   
309       2021-10-01           2021-10                   Unmapped   
..               ...               ...                        ...   
455       2021-10-31           2021-10  Extended Access Provision   
456       2021-10-31           2021-10           General Practice   
457       2021-10-31           2021-10                      Other   
458       2021-10-31           2021-10       Primary Care Network   
459       2021-10-31           2021-10                   Unmapped   

     count_of_appointments  
305                   7107  
306                1130646  
307                  22639  
308                  22974  
309                  56201  
..                     ...  
455                   1113  
456                  17449  
457                   1147  
458                   1200  
459                   2863  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
765       2022-01-01           2022-01  Extended Access Provision   
766       2022-01-01           2022-01           General Practice   
767       2022-01-01           2022-01                      Other   
768       2022-01-01           2022-01       Primary Care Network   
769       2022-01-01           2022-01                   Unmapped   
..               ...               ...                        ...   
915       2022-01-31           2022-01  Extended Access Provision   
916       2022-01-31           2022-01           General Practice   
917       2022-01-31           2022-01                      Other   
918       2022-01-31           2022-01       Primary Care Network   
919       2022-01-31           2022-01                   Unmapped   

     count_of_appointments  
765                    246  
766                   3578  
767                    280  
768                    215  
769                    572  
..                     ...  
915                   9962  
916                1305082  
917                  25034  
918                  30509  
919                  43240  

[155 rows x 4 columns]

appointment_date appointment_month            service_setting  \
1215       2022-04-01           2022-04  Extended Access Provision   
1216       2022-04-01           2022-04           General Practice   
1217       2022-04-01           2022-04                      Other   
1218       2022-04-01           2022-04       Primary Care Network   
1219       2022-04-01           2022-04                   Unmapped   
...               ...               ...                        ...   
1360       2022-04-30           2022-04  Extended Access Provision   
1361       2022-04-30           2022-04           General Practice   
1362       2022-04-30           2022-04                      Other   
1363       2022-04-30           2022-04       Primary Care Network   
1364       2022-04-30           2022-04                   Unmapped   

      count_of_appointments  
1215                   8011  
1216                1085037  
1217                  21077  
1218                  26493  
1219                  36836  
...                     ...  
1360                   3416  
1361                  23675  
1362                   1148  
1363                   3150  
1364                   1844  

[150 rows x 4 columns]

In [114]:
# Look at August 2021 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_summer, x = 'appointment_date', y = 'count_of_appointments', color = 'service_setting',
                 title = 'Total appointment across summer and service settings', 
                 labels = {'appointment_date': 'day',
                           'count_of_appointments': 'Number of appointments',
                           'service_setting': 'Service Setting'})
fig.show()

**Autumn (October 2021):**

In [115]:
# Look at October 2021 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_autumn, x = 'appointment_date', y = 'count_of_appointments', color = 'service_setting',
                 title = 'Total appointment across autumn and service settings', 
                 labels = {'appointment_date': 'day',
                           'count_of_appointments': 'Number of appointments',
                           'service_setting': 'Service Setting'})
fig.show()

**Winter (January 2022):**

In [116]:
# Look at January 2022 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_winter, x = 'appointment_date', y = 'count_of_appointments', color = 'service_setting',
                 title = 'Total appointment across winter and service settings', 
                 labels = {'appointment_date': 'day',
                           'count_of_appointments': 'Number of appointments',
                           'service_setting': 'Service Setting'})
fig.show()

**Spring (April 2022):**

In [117]:
# Look at April 2022 in more detail to allow a closer look.
# Create a lineplot.
fig = px.line(nc_ss_spring, x = 'appointment_date', y = 'count_of_appointments', color = 'service_setting',
                 title = 'Total appointment across spring and service settings', 
                 labels = {'appointment_date': 'day',
                           'count_of_appointments': 'Number of appointments',
                           'service_setting': 'Service Setting'})
fig.show()

# 

# Assignment activity 5

### Analyse tweets from Twitter with hashtags related to healthcare in the UK.

In [76]:
# Libraries and settings needed for analysis
import pandas as pd
import seaborn as sns

# Set figure size.
sns.set(rc={'figure.figsize':(15, 12)})

# Set the plot style as white.
sns.set_style('white')

# Maximum column width to display
pd.options.display.max_colwidth = 200

In [ ]:
# Load the data set.


# View the DataFrame.


In [ ]:
# Explore the metadata.


In [ ]:
# Explore the data set.


In [ ]:
# Would it be useful to only look at retweeted and favourite tweet messages?
# Explain your answer.


In [ ]:
# Create a new DataFrame containing only the text.


# View the DataFrame.


In [ ]:
# Loop through the messages, and create a list of values containing the # symbol.


In [ ]:
# Display the first 30 records.


In [ ]:
# Convert the series to a DataFrame in preparation for visualisation.


# Rename the columns.


In [ ]:
# Fix the count datatype.


# View the result.


In [ ]:
# Display records where the count is larger than 10.


In [ ]:
# Create a Seaborn barplot indicating records with a count >10 records.


In [ ]:
# Create the plot.


# View the barplot.


# 

# Assignment activity 6

### Investigate the main concerns posed by the NHS. 

In [ ]:
# Prepare your workstation.
# Load the appointments_regional.csv file.


# View the DataFrame.


In [ ]:
# Print the min and max dates.


In [ ]:
# Filter the data set to only look at data from 2021-08 onwards.


**Question 1:** Should the NHS start looking at increasing staff levels? 

In [ ]:
# Create an aggregated data set to review the different features.


# View the DataFrame.


In [ ]:
# Determine the total number of appointments per month.


# Add a new column to indicate the average utilisation of services.
# Monthly aggregate / 30 to get to a daily value.


# View the DataFrame.


In [ ]:
# Plot sum of count of monthly visits.
# Convert the appointment_month to string data type for ease of visualisation.


# Create a lineplot with Seaborn.


In [ ]:
# Plot monthly capacity utilisation.


# Create a lineplot.


**Question 2:** How do the healthcare professional types differ over time?

In [ ]:
# Create a line plot to answer the question.


**Question 3:** Are there significant changes in whether or not visits are attended?

In [ ]:
# Create a line plot to answer the question.


**Question 4:** Are there changes in terms of appointment type and the busiest months?

In [ ]:
# Create a line plot to answer the question.


**Question 5:** Are there any trends in time between booking an appointment?

In [ ]:
# Create a line plot to answer the question.


**Question 6:** How do the spread of service settings compare?

In [ ]:
# Let's go back to the national category DataFrame you created in an earlier assignment activity.


In [ ]:
# Create a new DataFrame consisting of the month of appointment and the number of appointments.

# View the DataFrame.


In [1]:
# Create a boxplot to investigate spread of service settings.


In [ ]:
# Create a boxplot to investigate the service settings without GP.


# 

### Provide a summary of your findings and recommendations based on the analysis.

> Double click to insert your summary.